iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -0,0 +1,317 @@
+  assert(input.getType().cast<VectorType>().getRank() == 1);
+  assert(dstVecType.getRank() == 1);
+  SmallVector<int64_t> offsets{position};
+  SmallVector<int64_t> strides{1};
+  return rewriter.create<vector::ExtractStridedSliceOp>(
+      loc, input, offsets, dstVecType.getShape(), strides);
+}
+
+// Helper to flatten a N-dimensional vector to a 1D vector.
+static Value flatten(PatternRewriter &rewriter, Location loc, Value vector) {
+  VectorType inputVecType = vector.getType().cast<VectorType>();
+  VectorType dstType = VectorType::get(inputVecType.getNumElements(),
+                                       inputVecType.getElementType());
+  return rewriter.create<vector::ShapeCastOp>(loc, dstType, vector);
+}
+
+/// Converts matrix-times-matrix-transposed vector.contracts with
+/// lhs and rhs inputs defined by arith.extsi promoting from i8 to i32,
+///
+///     %lhs_i32 = arith.extsi %lhs_i8 : i8 to i32
+///     %rhs_i32 = arith.extsi %rhs_i8 : i8 to i32
+///     %result = vector.contract [...]
+///                 %lhs_i32 : vector<8x4xi32>,
+///                 %rhs_i32 : vector<8x4xi32>,
+///                 %acc_i32 : vector<8x8xi32>,
+///                 [...]
+///
+/// To vector ops reading directly from the %lhs_i8 and %rhs_i8 values
+/// (bypassing the existing arith.extsi) and passing that to a llvm.inline_asm
+/// block implementing the matrix multiplication arithmetic using Aarch64
+/// dot-product instructions (sdot).
+struct MMT_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm
+    : OpRewritePattern<vector::ContractionOp> {
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractionOp,
+                                PatternRewriter &rewriter) const override {
+    // Check if `contractionOp` matches, and obtain the un-promoted i8 input
+    // LHS and RHS vectors, `lhsI8` and `rhsI8`.
+    if (!isMatrixTimesMatrixTransposed(contractionOp, 8, 4, 8)) {
+      return failure();
+    }
+    Type I8Type = rewriter.getIntegerType(8);
+    Type I32Type = rewriter.getIntegerType(32);
+    VectorType accType = contractionOp.acc().getType().cast<VectorType>();
+    if (accType.getElementType() != I32Type) {
+      return failure();
+    }
+    Value lhsI8 = getExtSIInput(I8Type, I32Type, contractionOp.lhs());
+    Value rhsI8 = getExtSIInput(I8Type, I32Type, contractionOp.rhs());
+    if (!lhsI8 || !rhsI8) return failure();
+
+    // `contractionOp` matches, start rewriting it. We only reference
+    // the `lhsI8` and `rhsI8` values obtained above as the inputs of the
+    // arith.extsi, so this rewrite will leave the existing arith.extsi without
+    // any user (unless something else was using them), so they may be
+    // removed by another transformation.
+    Location loc = contractionOp.getLoc();
+    // Flatten the inputs to 1D vectors.
+    Value flatLhsI8 = flatten(rewriter, loc, lhsI8);
```

Comment:
As we discussed, I would rather not do this as early on, and do this flattening later on a separate pass to lower to LLVM.

Reply:
As discussed back then (sorry about the long hiatus): flattening this here is a requirement imposed on us by the`LLVM::InlineAsmOp`, which can only take vector-register operands if they are 1D.

This is where Geoffrey's intrinsics code path may choose to diverge: plain LLVM ARM NEON intrinsics are 1D as well, but there is also a concept of 2D ARM NEON intrinsics in MLIR (being fleshed out as use cases arise).  So intrinsics code paths have a choice here, but inline-asm code paths don't (unless LLVM::InlineAsmOp gets relaxed to allow ND vectors with implicit flattening semantics, but this sort of fanciness runs arguably counter to the spirit of inline asm).